### PR TITLE
Performance tweaks

### PR DESCRIPTION
**Pull request by pjaspers**
This pulls in a couple of performance tweaks to optimize generating activity formula variables. This can be a hotspot in certain conditions (if a zone has a lot of org units, where lot is > 5, this can lead to an explosion of formulas. e.g. we had a case where we needed to generate 13068 formulas.

I took this as my benchmark:

``` ruby
      invoice_entity = Invoicing::InvoiceEntity.new(nil, invoicing_request, options)
      invoice_entity.instance_variable_set(:@pyramid, read_artefact("pyramid", "yml"))
      invoice_entity.instance_variable_set(:@datacompound, read_artefact("data-compound", "yml"))
      invoice_entity.instance_variable_set(:@orbf_project, read_artefact("project", "yml"))
      puts "After setup #{Time.now - time}"
      time = Time.now
        invoice_entity.call
      end
      puts "Call took #{Time.now - time}"
      time = Time.now
```

Against the `master` branch this took (I always took three runs, to have it be somehwat trustworthy)

```
Call took 30.00353
Call took 24.014524
Call took 24.685995
```

By using the SubstitutionBuilder (77bebe3) and mutating the result, instead of merging (
12437a8):
```
Call took 22.285549
Call took 21.277349
Call took 19.502618
```

adding caching the level pattern values and period facts (2161c31):

```
Call took 20.275272
Call took 19.35144
Call took 19.568146
```

adding caching for the states_substitutions (1355363)

```
Call took 18.802241
Call took 19.2533
Call took 18.246414
```

So it saved us 5 seconds on generating one invoice, without breaking anything existing.

## Self proof reading checklist

- [x] Did I run Pronto and fixed all warnings
- [x] Make sure all naming and code will remain understandable in 6 month by someone new to the project or by you.
- [x] Did I test the right thing?
- [x] Did I test corner cases, non happy path (defensive testing)?
- [x] Check that I used the ad-hoc patterns and created files accordingly
- [ ] Check code efficiency with record intensive project
- [x] Update documentation,readme accordingly

Thanks!

### CODE REVIEW

**Comment by pjaspers** [comment]
Do not merge yet, I'm running into issues when running the complete data_test on ORBF2.

**Comment by pjaspers** [comment]
I've had to revert this commit: https://github.com/BLSQ/orbf-rules_engine/pull/44/commits/13553634e8a3632b17cc79dbcc7b809bbbd1c95c the cache key I was using was not specific enough and the resulting formulas were incorrect.

**Comment by pjaspers** [comment]
What I like about this whole thing, is that the data tests in orbf2 did manage to catch that I broke some functionality (too agressively caching the states_substitutions), what I don't like is that they feel like they should be in this repo, instead of the orbf2 repo.

Capturing is part of orbf2, using the artefacts can be something that is run here instead.

(For now we don't have an issue with this, since we're mostly working on both at the same time, so it's easy to run the orbf2 tests to catch regressions in the rules engine, but it does feel wrong)

**Comment by pjaspers** [comment]
I've added support for constructing the pyramid from a snapshot, this is needed for blsq/orbf2#103

All the rest does increase performance, profiling will be in a separate PR.

**Comment by mestachs** [comment]
I'm less confident about the level_pattern_values(orgunit) cache. But I trust you and orbf2 verification specs.

File: lib/orbf/rules_engine/builders/activity_formula_variables_builder.rb [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # frozen_string_literal: true
-
+require_relative "./substitution_builder"
```
Comment (mestachs) [inline lib/orbf/rules_engine/builders/activity_formula_variables_builder.rb]: I would keep this in lib/orbf/rules_engine.rb (it's the only place where we do this)

**Comment by mestachs** [comment]
nice 1 second gain on my laptop for the test suite on my laptop compared to current master.
